kaldb/src/main/java/com/slack/kaldb/logstore/search/KaldbDistributedQueryService.java

```diff
@@ -44,39 +37,105 @@
   public static final KaldbSearch.SearchResult emptyResult =
       KaldbSearch.SearchResult.newBuilder().setFailedNodes(1).setTotalNodes(1).build();
 
-  // TODO Integrate with ZK to update list of servers
+  // For now we will use SearchMetadataStore to populate servers
+  // But this is wasteful since we add snapshots more often than we add/remove nodes ( hopefully )
+  // So this should be replaced cache/index metadata store when that info is present in ZK
+  // Whichever store we fetch the information in the future, we should also store the
+  // protocol that the node can be contacted by there since we hardcode it today
+  public KaldbDistributedQueryService(SearchMetadataStore searchMetadataStore) {
+    this.searchMetadataStore = searchMetadataStore;
+    this.searchMetadataStore.addListener(searchServersListener());
+    initStubs();
+  }
+
+  private void initStubs() {
+    List<SearchMetadata> searchMetadataNodes;
+    try {
+      searchMetadataNodes = searchMetadataStore.list().get();
+      searchMetadataNodes.forEach(
+          searchMetadata ->
+              stubs.put(
+                  searchMetadata.url, getKaldbServiceGrpcClient(protocol + searchMetadata.url)));
+    } catch (Exception e) {
+      LOG.error("Cannot initStubs stubs");
+      // throw runtime exception?
+    }
+  }
+
+  private KaldbMetadataStoreChangeListener searchServersListener() {
+    return () -> {
+      try {
+        Set<String> latestSearchServers = new HashSet<>();
+        searchMetadataStore
+            .list()
+            .get()
+            .forEach(searchMetadata -> latestSearchServers.add(searchMetadata.url));
+
+        // add new servers
+        latestSearchServers.forEach(
+            server -> {
+              if (!stubs.containsKey(server)) {
+                stubs.put(server, getKaldbServiceGrpcClient(protocol + server));
+              }
+            });
+
+        // invalidate old servers that no longer exist
+        stubs
```

Comment:
I think it would be valuable to log the number of stubs added and removed as part of a run. Further, adding a counter for the changes and a guage to count the number of stubs would be very valuable to monitor this queue. 

In addition, a debug level message for the changes would help debugging immensely.

Reply:
Added everything except for the gauge! The reason for not adding the gauge was the number of  URLs we query is different from the number of search metadata objects. So the guage might give misleading info?

Reply:
Could you explain how a gauge will help ? There can be many search requests being made and a gauge won't help tell us on a per search basis the cardinality ? 

I think tracing would be a better fit no?

Reply:
I think a guage would help us detect if too many stubs are added or removed. Tracing is not a great fit here since this logic is not in the request path.

I think for now let's just add how many stubs are added and removed on the log message on line 114.

Reply:
added